File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        def dfs(node: "TreeNode") -> "TreeNode":
            # If current node's value is greater than p's and q's value, go left
            if node.val > p.val and node.val > q.val:
                return dfs(node.left)

            # If current node's value is less than p's and q's value, go right
            if node.val < p.val and node.val < q.val:
                return dfs(node.right)

            # Return the lowest common ancestor
            return node

        # Return empty if any of root, p, and q is empty
        if not root or not p or not q:
            return None
        return dfs(root)

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        // Return empty if any of preorder and inorder is empty
        // Note that this is the base case of DFS, not an early return
        if (preorder.empty() || inorder.empty()) {
            return nullptr;
        }

        // Create root node using preorder[0]
        // because index 0 in preorder will always be the root of a subtree
        TreeNode* root = new TreeNode(preorder[0]);

        // Find preorder[0] in inorder and get preorder[0]'s index in inorder
        // Note that iteratorA - iteratorB = the distance between iteratorA and iteratorB
        // Ex.  iteratorA is a position 3, iteratorB is at position 0(begin()),
        //      iteratorA - iteratorB = 3 = iteratorA's index
        auto mid = find(inorder.begin(), inorder.end(), preorder[0]) - inorder.begin();

        // preorder = root | preLeft | preRight
        // inorder = inorderLeft | root | inorderLeft
        // Note that vector(_First, _Last) doesn't include _Last
        vector preLeft(preorder.begin() + 1, preorder.begin() + 1 + mid);
        vector preRight(preorder.begin() + 1 + mid, preorder.end());
        vector inLeft(inorder.begin(), inorder.begin() + mid);
        vector inRight(inorder.begin() + mid + 1, inorder.end());

        // Pass in preLeft, preRight, inLeft, inRight for recursive call
        root->left = buildTree(preLeft, inLeft);
        root->right = buildTree(preRight, inRight);
        return root;
    }
};
